### PR TITLE
Add clarification that GetFunctionPointerForDelegate honors UnmanagedFunctionPointerAttribute

### PR DESCRIPTION
**Pull request by PathogenDavid**
The [documentation for Marshal.GetFunctionPointerForDelegate](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getfunctionpointerfordelegate?view=netframework-4.7.2#System_Runtime_InteropServices_Marshal_GetFunctionPointerForDelegate__1___0_) states the following:

> The delegate `d` is converted to a function pointer that can be passed to unmanaged code by **using the [__stdcall](https://msdn.microsoft.com/library/e212594b-1827-4d07-9527-7d412b300df8) calling convention**.

(Emphasis mine)

This is incorrect/misleading. It will use the stdcall calling convention by _default_, but [experimentation](https://github.com/PathogenPlayground/DoesGetFunctionPointerForDelegateHonorConvention) with both .NET Framework and .NET Core reveals it will honor `UnmanagedFunctionPointerAttribute` if it is present on `TDelegate`.

This PR changes the wording for both overloads of `GetFunctionPointerForDelegate` to clarify that they honor `UnmanagedFunctionPointerAttribute` if it is present and use stdcall otherwise. I also updated the link to the `__stdcall` article from MSDN to the new Microsoft Docs article.

### CODE REVIEW

**Comment by mairaw** [comment]
Thanks for your contribution @PathogenDavid. @AaronRobinsonMSFT, @jkoritzinsky can one of you please review this change?

File: xml/System.Runtime.InteropServices/Marshal.xml [hunk 1]
```diff
@@ -3501,7 +3501,7 @@
           <format type="text/markdown"><![CDATA[  
   
 ## Remarks  
- The delegate `d` is converted to a function pointer that can be passed to unmanaged code using the [__stdcall](https://msdn.microsoft.com/library/e212594b-1827-4d07-9527-7d412b300df8) calling convention.  
+ The delegate `d` is converted to a function pointer that can be called by unmanaged code using the calling convention specified by the <xref:System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute> associated with the delegate's type or [__stdcall](https://docs.microsoft.com/cpp/cpp/stdcall) if none is specified.
```
Comment (mairaw) [inline xml/System.Runtime.InteropServices/Marshal.xml]:
```suggestion
 The delegate `d` is converted to a function pointer that can be called by unmanaged code using the calling convention specified by the <xref:System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute> associated with the delegate's type or [__stdcall](/cpp/cpp/stdcall) if none is specified.
```

File: xml/System.Runtime.InteropServices/Marshal.xml [hunk 2]
```diff
@@ -3563,7 +3563,7 @@
           <format type="text/markdown"><![CDATA[  
   
 ## Remarks  
- The delegate `d` is converted to a function pointer that can be passed to unmanaged code by using the [__stdcall](https://msdn.microsoft.com/library/e212594b-1827-4d07-9527-7d412b300df8) calling convention.  
+ The delegate `d` is converted to a function pointer that can be called by unmanaged code using the calling convention specified by the <xref:System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute> associated with `TDelegate` or [__stdcall](https://docs.microsoft.com/cpp/cpp/stdcall) if none is specified.
```
Comment (mairaw) [inline xml/System.Runtime.InteropServices/Marshal.xml]:
```suggestion
 The delegate `d` is converted to a function pointer that can be called by unmanaged code using the calling convention specified by the <xref:System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute> associated with `TDelegate` or [__stdcall](/cpp/cpp/stdcall) if none is specified.
```

**Comment by AaronRobinsonMSFT** [comment]
Thanks for the contribution @PathogenDavid. 

>This is incorrect/misleading.

I don't really see how this is incorrect or misleading, but rather both the default behavior as well as how to modify it should be documented. We should also clarify that the `__stdcall` is only for x86, not x64. As such the documentation should state the default case behavior first (i.e. `__stdcall` or `x64` calling convention) or the user can optionally specify an alternative using the `UnmanagedFunctionPointerAttribute` attribute.

File: xml/System.Runtime.InteropServices/Marshal.xml [hunk 3]
```diff
@@ -3501,7 +3501,7 @@
           <format type="text/markdown"><![CDATA[  
   
 ## Remarks  
- The delegate `d` is converted to a function pointer that can be passed to unmanaged code using the [__stdcall](https://msdn.microsoft.com/library/e212594b-1827-4d07-9527-7d412b300df8) calling convention.  
+ The delegate `d` is converted to a function pointer that can be called by unmanaged code using the calling convention specified by the <xref:System.Runtime.InteropServices.UnmanagedFunctionPointerAttribute> associated with the delegate's type or [__stdcall](/cpp/cpp/stdcall) if none is specified.
```
Comment (AaronRobinsonMSFT) [inline xml/System.Runtime.InteropServices/Marshal.xml]: Please indicate the default first and then the way to modify the default. Note that the usage of `__stdcall` is only for x86, the [x64 calling convention](https://docs.microsoft.com/cpp/build/x64-calling-convention) is used on x64 platform.

**Comment by PathogenDavid** [comment]
@mairaw 

Thanks! I didn't know you could do that. The style guide only showed relative links within the .NET documentation so I used a complete link to stay on the safe side.

---------

@AaronRobinsonMSFT 

> I don't really see how this is incorrect or misleading, but rather both the default behavior as well as how to modify it should be documented.

Misleading in the sense that it didn't make it clear that it was default behavior. The current wording makes it sound like it's the only behavior.

> We should also clarify that the __stdcall is only for x86, not x64.

I was thinking about this as I made my changes, actually. I felt like it was outside the scope of this change, and I wasn't sure how it applied on ARM platforms.

Additionally, isn't x64 calling convention not only the default on x64, but also the only option? I was under the impression that both `DllImportAttribute.CallingConvention` and `UnmanagedFunctionPointerAttribute.CallingConvention` were basically ignored on x64. If so, these remarks might start getting quite a bit longer to cover all of the scenarios.

-------------

This actually has me thinking that there should be a separate article that describes how calling conventions work on various platforms. (Or does one already exist?)

Something along the rough lines of:

* On x86, the following applies:
  * The default calling convention is `StdCall` (technically isn't it Winapi, which is generally Stdcall?)
  * The calling convention can be overridden by `DllImport.CallingConvention` or  `UnmanagedFunctionPointer`
* On x64 the following applies:
  * On Windows, the calling convention is always [the Microsoft x64 calling convention](https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention).
  * On Linux/macOS, the calling convention is always [the SystemV AMD64 ABI](https://en.wikipedia.org/wiki/X86_calling_conventions#System_V_AMD64_ABI). (Assuming that's true.)
  * On x64, the calling convention cannot be overriden.
* On ARM64...
* On ARM32...

Then this article can be linked from [DllImportAttribute.CallingConvention](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.dllimportattribute.callingconvention), the [UnmanagedFunctionPointerAttribute constructor](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.unmanagedfunctionpointerattribute.-ctor), and [Marshal.GetFunctionPointerForDelegate](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.getfunctionpointerfordelegate#System_Runtime_InteropServices_Marshal_GetFunctionPointerForDelegate__1___0_).

For example:

> The delegate `d` is converted to a function pointer that can be called by unmanaged code.
>
> The calling convention of the function is defined by **native calling conventions in .NET**.

(Where the bolded text is a link to said article.)

**Comment by AaronRobinsonMSFT** [comment]
This part of the code is very readable - https://github.com/dotnet/coreclr/blob/2929df03c2defe59381ccdaf7cec15ecde0e4140/src/vm/dllimport.cpp#L3251-L3300

**Comment by PathogenDavid** [comment]
>I am not aware of different calling conventions on ARM, but willing to be educated if there are.

I'm in the same boat. A quick glance at Wikipedia and ARM's compiler documentation seems to suggest there aren't any.

> Just realized you probably meant the enum value not the macro - sorry.

Yup, and no worries!

> I don't believe so, but @jkotas or @janvorli may know of one.

If neither of them know of an article that fits the bill, I will try to find some time to draft one.

> This part of the code is very readable

Thanks for that!

I was actually poking around in CoreCLR before I made my experiment program, but I originally started from the `GetFunctionPointerForDelegate` end of things and by the time I found `dllimport.cpp` I think I was too tired by then to pay much attention in it.

**Comment by tdykstra** [comment]
@PathogenDavid This issue has been dormant a while -- do you still plan to write an article? In the interim, would it make sense to limit the scope of this PR and just address the change request from the review, leaving the more comprehensive coverage for a separate PR later?

**Comment by PathogenDavid** [comment]
Yeah, sorry about that. I got busy around the time this issue last left off and I've been putting off getting back into the mindset to write the article. I'll try to find some time in the new week or so to get back up to speed on things and write something up.

**Comment by carlossanlop** [comment]
> _In the interim, would it make sense to limit the scope of this PR and just address the change request from the review, leaving the more comprehensive coverage for a separate PR later?_

@PathogenDavid would you mind doing @tdykstra 's suggestion so we can get this change merged?

**Comment by tdykstra** [comment]
@AaronRobinsonMSFT It's been over a month with no response from @PathogenDavid  -- Would it be better to close the PR or merge it without the changes you requested?

**Comment by AaronRobinsonMSFT** [comment]
@tdykstra Sure. If @PathogenDavid wants, he can resubmit the changes.

**Comment by PathogenDavid** [comment]
Sorry for letting this fall off my radar again. It appears that #3706 and #3265 cover the changes this PR originally set out to accomplish, so this PR should've been closed anyway.

I hope revisit the idea of a more fleshed-out article in the future, assuming it doesn't end up happening by default as part of https://github.com/dotnet/docs/issues/11396 before then.

Sorry again for the inconvenience.

**Pull request closed without merging**